internal/controllers/topology/cluster/reconcile_state_test.go

```diff
@@ -269,6 +276,583 @@ func TestReconcileShim(t *testing.T) {
+						},
+					},
+				},
+			},
+			hookResponse:       failureResponse,
+			wantMarked:         true,
+			wantHookToBeCalled: true,
+			wantError:          true,
+		},
+		{
+			name: "hook should not be called if it is marked and the control plane is not ready - the hook should remain marked",
+			cluster: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-cluster",
+					Namespace: "test-ns",
+					Annotations: map[string]string{
+						runtimev1.PendingHooksAnnotation: "AfterControlPlaneInitialized",
+					},
+				},
+				Spec: clusterv1.ClusterSpec{
+					ControlPlaneRef:   &corev1.ObjectReference{},
+					InfrastructureRef: &corev1.ObjectReference{},
+				},
+				Status: clusterv1.ClusterStatus{
+					Conditions: clusterv1.Conditions{
+						clusterv1.Condition{
+							Type:   clusterv1.ControlPlaneInitializedCondition,
+							Status: corev1.ConditionFalse,
+						},
+					},
+				},
+			},
+			hookResponse:       failureResponse,
+			wantMarked:         true,
+			wantHookToBeCalled: false,
+			wantError:          false,
+		},
+		{
+			name: "hook should not be called if it is not marked",
+			cluster: &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-cluster",
+					Namespace: "test-ns",
+				},
+				Spec: clusterv1.ClusterSpec{
+					ControlPlaneRef:   &corev1.ObjectReference{},
+					InfrastructureRef: &corev1.ObjectReference{},
+				},
+				Status: clusterv1.ClusterStatus{
+					Conditions: clusterv1.Conditions{
+						clusterv1.Condition{
+							Type:   clusterv1.ControlPlaneInitializedCondition,
+							Status: corev1.ConditionTrue,
+						},
+					},
+				},
+			},
+			hookResponse:       failureResponse,
+			wantMarked:         false,
+			wantHookToBeCalled: false,
+			wantError:          false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			s := &scope.Scope{
+				Current: &scope.ClusterState{
+					Cluster: tt.cluster,
+				},
+				HookResponseTracker: scope.NewHookResponseTracker(),
+			}
+
+			fakeRuntimeClient := fakeruntimeclient.NewRuntimeClientBuilder().
+				WithCallAllExtensionResponses(map[runtimecatalog.GroupVersionHook]runtimehooksv1.ResponseObject{
+					afterControlPlaneInitializedGVH: tt.hookResponse,
+				}).
+				WithCatalog(catalog).
+				Build()
+
+			fakeClient := fake.NewClientBuilder().WithObjects(tt.cluster).Build()
+
+			r := &Reconciler{
+				Client:        fakeClient,
+				APIReader:     fakeClient,
+				RuntimeClient: fakeRuntimeClient,
+			}
+
+			err := r.callAfterControlPlaneInitialized(ctx, s)
+			g.Expect(fakeRuntimeClient.CallAllCount(runtimehooksv1.AfterControlPlaneInitialized) == 1).To(Equal(tt.wantHookToBeCalled))
+			g.Expect(hooks.IsPending(runtimehooksv1.AfterControlPlaneInitialized, tt.cluster)).To(Equal(tt.wantMarked))
+			g.Expect(err != nil).To(Equal(tt.wantError))
+		})
+	}
+}
+
+func TestReconcile_callAfterClusterUpgrade(t *testing.T) {
```

Comment:
nit: If we introduce isUpgradCompleted as suggested above this test can be eventually split in two/simplified, but no objections to keep it as it is